### ISSUE TITLE
Peak Tracker never finishes

### ISSUE THREAD

**Issue by karlduderstadt** [duderstadt-lab/mars-core#21]
Sometime when running the PeakTracker on our linux cluster the command never finished. The progress bar goes to the end and then remains there and the command never finishes. This appears to only happen when using a fit radius of 4 and only for certain videos. 

It looks like there might be an infinite loop created during fitting and this problem only occur under very specific conditions. Perhaps a certain type of peak near a border. Probably one frame gets this unique peaks and triggers the infinite loop. Then the progress bar goes to the end except for that one frame, where it just remains.

Need to try and reproduce and investigate further and update the fitting routine to prevent this infinite loop

**Comment by karlduderstadt** [comment]
This has happened to me, but in the last week, I have not been able to reproduce the issue. Please if someone discovers a video that does this consistently it would be a huge help if you could give it to me for further testing.

I have noticed this happens on our linux HPC where sometimes the progress bars goes to the end and then nothing happens after that, the command stalls. However, I had two cases this week, where simply closing and restarting the same command resulted in no problems. So maybe this means it is some kind of multithreading problem? Race condition or just a problem with linux HPC. It seems to happen rarely.

**Comment by mjosch** [comment]
I kept having the same issue in the past and just again a few days ago (only on our linux HPC). Unfortunately, as you already noticed, restarting the exact same command on the same video solves the problem, so far, I haven't found a video which consistently produces this problem.

**Comment by karlduderstadt** [comment]
@mjosch thanks for the feedback! Ok well I guess at least it is consistent. Each thread will retrieve an image and work on it. I wonder if somehow a thread sometimes fails to retrieve the file on HPC and then stalls. Maybe this would explain why I don't have this issue on my mac. I am just not sure how to troubleshoot. I will keep an eye out for a dataset and causes this issue always.

There was another issue in the R2 calculation that caused infinite loop during peak fitting but that was fixed a while ago. I thought this issue might be related, but then it should happen every time. Just for reference that issue was:
https://github.com/duderstadt-lab/mars-core/commit/6e92c5fff2d5c05446df9e9841208bbddd12568f#diff-6b0f4bc683185e3d5c4634c06e873bbc8fee76e26aebac50bcd9c764221f0c6d

I will keep testing and see if I somehow get a more solid solution. In the meantime, I will leave the issue open.

**Comment by karlduderstadt** [comment]
I think this is related to the way multithreading works with scijava ops. I recently discovered that image ops also perform multithreading. In our image processing commands, I usually multithread on images, so several threads are working on individual images at the same time. Unfortunately, when I run ops on an image that also creates a bunch of threads, so it turns out that the total threads is a lot: my threads * ops threads. I believe sometimes the creation of a very large number of threads causing the stalling of commands.

Recently, working with Anita, we switch to using one thread in our dialogs and it solved this problem. I also discussed this with the scijava developers and very kindly they add a method to the ops framework to set the thread count. This also seems to help, however, this is not yet in Fiji. When that moves to Fiji, I will set the threads for ops to 1 and when we can use what ever thread count we like and I hope that address this issue once and for all.

I could also be wrong and it is something else, but this thread conflict fits with it only happening sometimes and not always, even with the same settings.
